contrib/udfs/src/test/java/org/apache/drill/exec/udfs/TestUserAgentFunctions.java

```diff
@@ -0,0 +1,90 @@
+      "t1.ua.DeviceBrand AS DeviceBrand,\n" +
+      "t1.ua.DeviceCpuBits AS DeviceCpuBits,\n" +
+      "t1.ua.OperatingSystemClass AS OperatingSystemClass,\n" +
+      "t1.ua.OperatingSystemName AS OperatingSystemName,\n" +
+      "t1.ua.OperatingSystemVersion AS OperatingSystemVersion,\n" +
+      "t1.ua.OperatingSystemVersionMajor AS OperatingSystemVersionMajor,\n" +
+      "t1.ua.OperatingSystemNameVersion AS OperatingSystemNameVersion,\n" +
+      "t1.ua.OperatingSystemNameVersionMajor AS OperatingSystemNameVersionMajor,\n" +
+      "t1.ua.LayoutEngineClass AS LayoutEngineClass,\n" +
+      "t1.ua.LayoutEngineName AS LayoutEngineName,\n" +
+      "t1.ua.LayoutEngineVersion AS LayoutEngineVersion,\n" +
+      "t1.ua.LayoutEngineVersionMajor AS LayoutEngineVersionMajor,\n" +
+      "t1.ua.LayoutEngineNameVersion AS LayoutEngineNameVersion,\n" +
+      "t1.ua.LayoutEngineBuild AS LayoutEngineBuild,\n" +
+      "t1.ua.AgentClass AS AgentClass,\n" +
+      "t1.ua.AgentName AS AgentName,\n" +
+      "t1.ua.AgentVersion AS AgentVersion,\n" +
+      "t1.ua.AgentVersionMajor AS AgentVersionMajor,\n" +
+      "t1.ua.AgentNameVersionMajor AS AgentNameVersionMajor,\n" +
+      "t1.ua.AgentLanguage AS AgentLanguage,\n" +
+      "t1.ua.AgentLanguageCode AS AgentLanguageCode,\n" +
+      "t1.ua.AgentSecurity AS AgentSecurity\n" +
+      "FROM (SELECT parse_user_agent('Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US; rv:1.8.1.11) Gecko/20071127 Firefox/2.0.0.11') AS ua FROM (values(1))) AS t1";
+
+    testBuilder().sqlQuery(query).unOrdered()
+      .baselineColumns("DeviceClass", "DeviceName", "DeviceBrand","DeviceCpuBits","OperatingSystemClass", "OperatingSystemName","OperatingSystemVersion",
+        "OperatingSystemVersionMajor","OperatingSystemNameVersion","OperatingSystemNameVersionMajor","LayoutEngineClass","LayoutEngineName","LayoutEngineVersion",
+        "LayoutEngineVersionMajor","LayoutEngineNameVersion","LayoutEngineBuild","AgentClass","AgentName","AgentVersion","AgentVersionMajor","AgentNameVersionMajor",
+        "AgentLanguage","AgentLanguageCode","AgentSecurity")
+      .baselineValues("Desktop","Desktop", "Unknown","32", "Desktop", "Windows NT", "XP", "XP", "Windows XP", "Windows XP", "Browser", "Gecko", "1.8.1.11", "1", "Gecko 1.8.1.11",
+        "20071127", "Browser", "Firefox", "2.0.0.11", "2", "Firefox 2", "English (United States)", "en-us","Strong security")
+      .go();
+  }
+
+  @Test
+  public void testGetHostName() throws Exception {
+    String query = "SELECT parse_user_agent('Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US; rv:1.8.1.11) Gecko/20071127 Firefox/2.0.0.11', 'AgentSecurity') AS agent FROM " +
+      "(values" +
+      "(1))";
+    testBuilder().sqlQuery(query).ordered().baselineColumns("agent").baselineValues("Strong security").go();
+
+    query = "SELECT parse_user_agent('Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US; rv:1.8.1.11) Gecko/20071127 Firefox/2.0.0.11', '') AS agent FROM " +
```

Comment:
1. Please split into two tests.
2. What will happen if user passed null for the first / second / both parameters? Drill will return null? Please add tests for these cases as well.

Reply:
@arina-ielchiieva 
The issue here is the function cannot accept null as an input.  Any function which has a `ComplexWriter` as output will not accept nullable holders as input.  Hence, the issue.  If there's a way to resolve this, I'd love to know, but I put a note in the documentation about this.

Reply:
@cgivre have you tried having `@Param NullableVarCharHolder input` and passing `CAST(null as VARCHAR)` (i.e., specifying the correct type of input) for `NULL` input value in case of `@Output BaseWriter.ComplexWriter outWriter`. Is it still not accepted?

Reply:
@KazydubB 
If you put any kind of Nullable holder as input, Drill doesn't recognize the UDF at all.  Believe me, I've tried...

Reply:
To add to that, I believe this issue with the `ComplexWriter` should be fixed such that UDFs which have complex output should be able to accept `null` input and do something other than throw exceptions.

Reply:
@cgivre there is at least one such function `parse_query` (`ParseQueryFunction.ParseQueryNullableInput.java`) which has: